cartoframes/io/managers/gbq_manager.py

```diff
@@ -0,0 +1,176 @@
+        bounds = result.iloc[0]
+        zoom = math.floor(math.log2(360 / (bounds.xmax - bounds.xmin)))
+        return [[bounds.xmin, bounds.ymin], [bounds.xmax, bounds.ymax]], zoom
+
+    def trigger_mvt_generation(self, query, zoom):
+        table_name = create_hash(query)
+
+        if self.check_table_exists(table_name):
+            log.info('DEBUG: table cached')
+            return
+
+        xo = 360./(2**zoom)
+        yo = 180./(2**zoom)
+
+        # TODO: optimize query
+        generation_query = '''
+        CREATE TABLE {dataset}.{table} AS (
+            WITH data AS (
+                {query}
+            ),
+            data_bounds AS (
+                SELECT geoid, rmr_tests.ST_Envelope_Box(TO_HEX(ST_ASBINARY(geom))) AS bbox
+                FROM data
+            ),
+            global_bounds AS (
+                SELECT
+                    MIN(bbox[OFFSET(0)]) as gxmin,
+                    MAX(bbox[OFFSET(1)]) as gxmax,
+                    MIN(bbox[OFFSET(2)]) as gymin,
+                    MAX(bbox[OFFSET(3)]) as gymax
+                FROM data_bounds
+            ),
+            global_bbox AS (
+                SELECT tiler.getTilesBBOX(gxmin-{xo}, gymin-{yo}, gxmax+{xo}, gymax+{yo}, {zoom}, 16/4096) AS gbbox
+                FROM global_bounds
+            ),
+            tiles_bbox AS (
+                SELECT z, x, y, xmin, ymin, xmax, ymax
+                FROM global_bbox
+                CROSS JOIN UNNEST(global_bbox.gbbox)
+            ),
+            tiles_xyz AS (
+                SELECT b.z, b.x, b.y, a.geoid
+                FROM data_bounds a, tiles_bbox b
+                WHERE NOT ((bbox[OFFSET(0)] > b.xmax) OR
+                           (bbox[OFFSET(1)] < b.xmin) OR
+                           (bbox[OFFSET(2)] > b.ymax) OR
+                           (bbox[OFFSET(3)] < b.ymin))
+            ),
+            tiles_geom AS (
+                SELECT b.z, b.x, b.y, a.geoid, ST_ASGEOJSON(a.geom) AS geom, a.* EXCEPT (geoid, geom)
+                FROM data a, tiles_xyz b
+                WHERE a.geoid = b.geoid
+            ),
+            tiles_mvt AS (
+                SELECT tiler.ST_ASMVT(b.z, b.x, b.y, ARRAY_AGG(TO_JSON_STRING(a)), 0) AS tile
+                FROM tiles_geom a, tiles_xyz b
+                WHERE a.geoid = b.geoid AND a.x = b.x AND a.y = b.y AND a.z = b.z
+                GROUP BY b.z, b.x, b.y
+            )
+            SELECT z, x, y, mvt
+            FROM tiles_mvt
+            CROSS JOIN UNNEST(tiles_mvt.tile)
+        )
+        '''.format(dataset=MVT_DATASET, table=table_name, query=query,
+                   xo=xo, yo=yo, zoom=zoom)
+        job = self.client.query(generation_query)
+        job.result()  # Wait for the job to complete.
+
+    def estimated_data_size(self, query):
+        log.info('Estimating size. This may take a few seconds')
+        estimation_query = '''
+            WITH q as ({})
+            SELECT SUM(CHAR_LENGTH(ST_ASTEXT(geom))) AS s FROM q
```

Comment:
I think it requires to run the full query twice. If the query is expensive it will be a problem

Reply:
We are running the query more than twice. It's something we must tackle to optimize. More info here: https://app.clubhouse.io/cartoteam/story/59018/api-design-cf-interface#activity-60585.